marshalable-derive/src/lib.rs

```diff
@@ -65,25 +94,271 @@ fn derive_tpm_marshal_inner(input: DeriveInput) -> Result<TokenStream> {
+                fn try_unmarshal(buffer: &mut tpm2_rs_marshalable::UnmarshalBuf) -> tpm2_rs_marshalable::exports::errors::TpmRcResult<Self> {
+                    #unmarsh_text
+                }
 
-            fn try_marshal(&self, buffer: &mut [u8]) -> tpm2_rs_marshalable::exports::errors::TpmRcResult<usize> {
-                let mut written: usize = 0;
-                #marsh_text;
-                Ok(written)
+                fn try_marshal(&self, buffer: &mut [u8]) -> tpm2_rs_marshalable::exports::errors::TpmRcResult<usize> {
+                    let mut written: usize = 0;
+                    #marsh_text;
+                    Ok(written)
+                }
             }
         }
     };
-
     Ok(expanded)
 }
 
+fn derive_tpm2b_code(
+    tpm2b_outer_struct_name: &Ident,
+    stru: &syn::DataStruct,
+    tpm2b_simple: bool,
+    tpm2b_struct: bool,
+) -> TokenStream {
+    const TPM2B_SIMPLE_ERR_GENERAL: &str = "A Tpm2b struct must contain just to elements, a u16 size and a buffer array when using #[marshalable(tpm2b_simple)]";
+    const TPM2B_SIMPLE_ERR_1ST_ELM: &str = "First element in Tpm2b struct must be a 'size: u16' when using #[marshalable(tpm2b_simple)]";
+    const TPM2B_SIMPLE_ERR_2ND_ELM: &str = "Second element in Tpm2b struct must be u8 buffer: <buffer name>: [u8, <buffer size>] when using #[marshalable(tpm2b_simple)].";
+    const TPM2B_STRUCT_ERR_2ND_ELM: &str = "Second element in Tpm2b struct must define a u8 buffer using sizeof expression: <buffer name>: [u8, sizeof( <some struct>) ] when using #[marshalable(tpm2b_struct)].";
+
+    // Note: At least one of "tpm2b_simple" and "tpm2b_struct" is true.
+    assert!(
+        tpm2b_simple,
+        "We can't have #[marshalable(tpm2b_struct)] without #[marshalable(tpm2b_simple)]"
+    );
+
+    if let Fields::Named(ref fields) = stru.fields {
+        // First make sure we have precisely 2 elements:
+        assert!(fields.named.len() == 2, "{}", TPM2B_SIMPLE_ERR_GENERAL);
+
+        // Lets validate the first element is named size and has type u16.
+        {
+            let first = &fields.named[0];
+            let field_name = first.ident.as_ref().unwrap();
+            let field_type = &first.ty;
+            assert!(
+                field_name == "size"
+                    && matches!(
+                        field_type,
+                        syn::Type::Path(type_path) if type_path.path.is_ident("u16")
+                    ),
+                "{}",
+                TPM2B_SIMPLE_ERR_1ST_ELM
+            );
+        };
+
+        // Lets validate the second element in the Tpm2b structure and extract name and size expression.
+        let second = &fields.named[1];
+        let field_name = second.ident.as_ref().unwrap();
+        let field_type = &second.ty;
+
+        // Expect array type.
+        let type_array = if let syn::Type::Array(type_array) = field_type {
+            type_array
+        } else {
+            panic!("{}", TPM2B_SIMPLE_ERR_2ND_ELM);
```

Comment:
please do not panic, instead propegate errors to top level similar to https://github.com/tpm-rs/tpm-rs/blob/bd995111e44f04bfbd677080d887672feebf60a7/marshalable-derive/src/lib.rs#L62

Reply:
Fixed